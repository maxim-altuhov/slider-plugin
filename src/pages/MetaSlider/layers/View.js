import ViewSlider from './ViewSlider';
import ViewScale from './ViewScale';
import ViewMarkers from './ViewMarkers';
import ViewThumbs from './ViewThumbs';
import ViewMinAndMaxValues from './ViewMinAndMaxValues';
import ViewError from './ViewError';

class View {
  constructor() {
    this.views = {
      viewError: new ViewError(),
      viewSlider: new ViewSlider(),
      viewThumbs: new ViewThumbs(),
      viewMarkers: new ViewMarkers(),
      viewMinAndMax: new ViewMinAndMaxValues(),
      viewScale: new ViewScale(),
    };
  }

  init(options) {
    Object.keys(this.views).forEach((view) => {
      if ('init' in this.views[view]) this.views[view].init(options);
    });
  }

  renderSlider(initSelector) {
    this.views.viewSlider.renderSlider(initSelector);
  }
}

export default View;
