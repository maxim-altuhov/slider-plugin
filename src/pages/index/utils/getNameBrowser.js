function getNameBrowser() {
  const ua = navigator.userAgent;

  if (ua.search(/rv:11.0/) > 0) {
    return 'Internet Explorer 11';
  }
  if (ua.search(/MSIE/) > 0) {
    return 'Internet Explorer';
  }
  return 'ะะต Internet Explorer';
}

export default getNameBrowser;
