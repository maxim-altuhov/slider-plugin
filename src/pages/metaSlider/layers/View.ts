import ViewSlider from './ViewSlider';
import ViewScale from './ViewScale';
import ViewMarkers from './ViewMarkers';
import ViewThumbs from './ViewThumbs';
import ViewError from './ViewError';

class View {
  /**
   * I use the any type to be able to access any method
   * of the class instance in the viewList object.
   */
  readonly viewList: Record<string, any> = {
    viewError: new ViewError(),
    viewScale: new ViewScale(),
    viewSlider: new ViewSlider(),
    viewThumbs: new ViewThumbs(),
    viewMarkers: new ViewMarkers(),
  };

  update(options: IPluginOptions) {
    Object.keys(this.viewList).forEach((view) => {
      if ('update' in this.viewList[view]) this.viewList[view].update(options);
    });
  }

  // Initial render of the slider and its main elements
  renderSlider($initSelector: JQuery) {
    this.viewList['viewSlider'].renderSlider($initSelector);
  }
}

export default View;
