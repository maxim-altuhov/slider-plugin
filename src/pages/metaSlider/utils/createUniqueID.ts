function createUniqueID(): string {
  const IDPartOne = Date.now().toString(36);
  const IDPartTwo = Math.random().toString(36).substring(2);

  return IDPartOne + IDPartTwo;
}

export default createUniqueID;
