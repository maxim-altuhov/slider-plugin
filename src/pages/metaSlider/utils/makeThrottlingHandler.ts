// The any type is used, since the called functions can contain any arguments
function makeThrottlingHandler(fn: (...args: any[]) => unknown, timeout: number) {
  let timer: NodeJS.Timeout | null = null;

  return (...args: any[]) => {
    if (timer) return;

    timer = setTimeout(() => {
      fn(...args);

      if (timer) {
        clearTimeout(timer);
        timer = null;
      }
    }, timeout);
  };
}

export default makeThrottlingHandler;
