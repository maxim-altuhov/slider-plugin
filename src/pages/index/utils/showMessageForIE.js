import getNameBrowser from './getNameBrowser';

function showMessageForIE() {
  const currentBrowser = getNameBrowser();
  const thisIsIEBrowser = (currentBrowser === 'Internet Explorer 11') || (currentBrowser === 'Internet Explorer');

  if (thisIsIEBrowser) {
    document.body.innerHTML = 'Internet Explorer is no longer supported, please switch to a more modern browser!';
  }
}

export default showMessageForIE;
