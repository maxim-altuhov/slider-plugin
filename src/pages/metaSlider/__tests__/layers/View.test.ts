/* eslint-disable @typescript-eslint/dot-notation */
import View from '../../layers/View';
import InitSettings from '../../data/InitSettings';
import * as createUniqueID from '../../utils/createUniqueID';

const initSelectorName = 'render-selector';
const initHTMLBlock = `<div id="${initSelectorName}"></div>`;
document.body.innerHTML = initHTMLBlock;
const $initSelector = $(`#${initSelectorName}`);

describe('Checking the "View" layer', () => {
  const classView = new View();

  Object.keys(classView.subViewsList).forEach((view) => {
    classView.subViewsList[view].update = jest.fn();
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  test('Checking the "renderSlider" method', () => {
    const mockCreateUniqueID = jest.spyOn(createUniqueID, 'default').mockImplementation(() => '');
    const checkingSelectorsArr = [
      /js-meta-slider/,
      /js-meta-slider__progress/,
      /js-meta-slider__thumb/,
      /js-meta-slider__marker/,
      /js-meta-slider__scale/,
      /data-id/,
      /data-value/,
      /data-text/,
    ];

    classView.renderSlider($initSelector);

    expect(document.body.innerHTML).toMatchSnapshot();
    checkingSelectorsArr.forEach((selector) => {
      expect(document.body.innerHTML).toMatch(selector);
    });
    expect(mockCreateUniqueID).toHaveBeenCalled();
    expect(document.body.innerHTML).not.toBe(initHTMLBlock);
    expect(classView['_$selector']).toHaveLength(1);
    expect(classView['_$selector']).toBe($initSelector);
  });

  test('When the "update" method is performed at the "View" level, an update occurs in the dependent subview from "subViewsList"-> %s and an object with options is passed to the "update"', () => {
    classView.updateViews(InitSettings);

    Object.keys(classView.subViewsList).forEach((view) => {
      expect(classView.subViewsList[view].update).toHaveBeenCalledWith(InitSettings);
    });
  });

  test('Checking the "updateModel" method', () => {
    const FAKE_TARGET_VALUE = 50;
    const FAKE_EVENT: Event = $.Event('click');
    jest.spyOn(classView, 'notify').mockImplementation();

    classView.updateModel(FAKE_EVENT, FAKE_TARGET_VALUE);

    expect(classView.notify).toHaveBeenCalledWith(FAKE_EVENT, FAKE_TARGET_VALUE);
  });
});
