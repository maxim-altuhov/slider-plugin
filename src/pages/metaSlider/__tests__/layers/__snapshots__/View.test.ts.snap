// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checking the "View" layer Checking the "renderSlider" method 1`] = `
"<div id=\\"render-selector\\" data-id=\\"\\"><div class=\\"meta-slider js-meta-slider\\"><div class=\\"meta-slider__progress js-meta-slider__progress\\"></div>
    <button type=\\"button\\" class=\\"meta-slider__thumb js-meta-slider__thumb meta-slider__thumb_left\\" data-value=\\"\\" data-text=\\"\\">
      <span class=\\"meta-slider__marker js-meta-slider__marker meta-slider__marker_left\\"></span>
    </button>
    <button type=\\"button\\" class=\\"meta-slider__thumb js-meta-slider__thumb meta-slider__thumb_right\\" data-value=\\"\\" data-text=\\"\\">
      <span class=\\"meta-slider__marker js-meta-slider__marker meta-slider__marker_right\\"></span>
    </button>
    <div class=\\"meta-slider__scale js-meta-slider__scale\\">
      <button type=\\"button\\" class=\\"meta-slider__scale-point js-meta-slider__scale-point\\" style=\\"display: none;\\"></button>
    </div></div></div>"
`;
